recipes-extended/xen/blktap3/fix-encryption.patch

```diff
@@ -0,0 +1,169 @@
+    Firstly, we always want to load the crypto support, so drop the dummy_load
+    routine. Secondly, we use tap-ctl-create, which was defaulting to NULL and 
+    0 for encryption and keysize since -E is only defined for tap-ctl-open.
+    This is also addressed by dropping the dummy load and if check
+ 4. findkey() sets keysize to 512 or 256, but an xts case deals with bytes,
+    so we divide by 8 before xts_aes_setkey()
+
+################################################################################
+CHANGELOG
+################################################################################
+Authors:
+Chris Rogers <rogersc@ainfosec.com>
+
+################################################################################
+REMOVAL
+################################################################################
+When upstreamed.
+
+################################################################################
+UPSTREAM PLAN
+################################################################################
+Should be upstreamed back to xapi project, but will take more time to update our
+implementation to fit their design changes. This needs to be in our tree ASAP
+so we patch here first, and upstream later.
+
+################################################################################
+INTERNAL DEPENDENCIES
+################################################################################
+
+################################################################################
+PATCHES
+################################################################################
+--- a/drivers/block-crypto.c
++++ b/drivers/block-crypto.c
+@@ -344,7 +344,7 @@ out:
+ #endif
+ 
+ int
+-vhd_open_crypto(vhd_context_t *vhd, const uint8_t *key, size_t key_bytes, const char *name)
++vhd_open_crypto(vhd_context_t *vhd, const uint8_t *keyin, size_t key_bytes, const char *name)
+ {
+ 	struct vhd_keyhash keyhash;
+ 	int err;
+@@ -402,7 +402,7 @@ vhd_open_crypto(vhd_context_t *vhd, cons
+ 		return err;
+ 	}
+ 
+-	xts_aes_setkey(vhd->xts_tfm, key, key_bytes);
++	xts_aes_setkey(vhd->xts_tfm, key, keysize / 8);
+ 	return 0;
+ }
+ 
+--- a/drivers/block-vhd.c
++++ b/drivers/block-vhd.c
+@@ -79,7 +79,7 @@
+ 
+ unsigned int SPB;
+ 
+-#define LIBBLOCKCRYPTO_NAME "libblockcrypto.so"
++#define LIBBLOCKCRYPTO_NAME "libblockcrypto.so.0"
+ 
+ #define DEBUGGING   2
+ #define MICROSOFT_COMPAT
+@@ -674,18 +674,6 @@ vhd_log_open(struct vhd_state *s)
+ 		allocated, full, s->next_db);
+ }
+ 
+-static int dummy_open_crypto(
+-	vhd_context_t *vhd, const uint8_t *key, size_t key_bytes,
+-	const char *name)
+-{
+-	if (key) {
+-		EPRINTF("Encryption requested with no support library\n");
+-		return -EINVAL;
+-	}
+-
+-	return 0;
+-}
+-
+ static int
+ __load_crypto(struct td_vbd_encryption *encryption)
+ {
+@@ -695,41 +683,35 @@ __load_crypto(struct td_vbd_encryption *
+ 		return -ENOMEM;
+ 	}
+ 
+-	if (encryption->encryption_key == NULL) {
```

Comment:
You could make a smaller patch by changing this to just:
```
-	if (encryption->encryption_key == NULL) {
+	if (0 /* OXT - force else path to always load encryption support */) {
```
and not changing the indenting?

Reply:
100% correct.  ```If (0)``` always seemed hacky to me but I can absolutely change it if we want a smaller patch.

Reply:
Can the patch work with the #ifdef?
Also, no value in removing dummy_open_crypto is there?

Thinking:
load_crypto = encryption->encryption_key != NULL
#ifdef OPEN_XT
load_crypto = 1
#endif

Reply:
Yes that is do-able.  I will make the changes and test.